### PR TITLE
randomize plugin added

### PR DESCRIPTION
**Pull request by namjohn920**
randomize plugin added to wet-boew.
It is added under provisional. 
The example is shown on provisional page as well.

### CODE REVIEW

File: site/pages/docs/ref/provisional-en.hbs [hunk 1]
```diff
@@ -111,3 +117,18 @@
 		&lt;p&gt;Secondary title&lt;/p&gt;
 	&lt;/div&gt;
 &lt;/div&gt;</code></pre>
+
+<h3 id="well-header-rwd-bgimg">Showing random image using [data-wb-randomize] plugin</h3>
+<p>Below example shows one image out of three images randomly. Refresh the page to see that the image changes</p>
+<div class="provisional" data-wb-randomize="toggle@self!img@{&quot;number&quot;:1,&quot;classes&quot;: &quot;hide&quot;}">
```
Comment (duboisp) [inline site/pages/docs/ref/provisional-en.hbs]: Can you review how the content in the attribute "data-wb-randomize" is structured. In WET4 we have either one single value or a JSON object that contains multiple values.

**Review by duboisp — CHANGES_REQUESTED** [review]

File: site/pages/docs/ref/provisional-en.hbs [hunk 2]
```diff
@@ -111,3 +117,18 @@
...
+	<img class="provisional-img-example hide" src="../../demos/tabs/img/investinourfuture.jpg" alt="Invest in our Future" />
+	<img class="provisional-img-example hide" src="../../demos/tabs/img/flytheskies.jpg" alt="Fly the Skies" />
+	<img class="provisional-img-example hide" src="../../demos/tabs/img/protect-environment.jpg" alt="Protect Environment" />
+</div>
+<br>
```
Comment (namjohn920) [inline site/pages/docs/ref/provisional-en.hbs]: get rid of br

File: src/plugins/wb-randomize/wb-randomize.js [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+ * @author @masterbee
+ */
+
+( function( $, wb ) {
+"use strict";
+
+$( ".provisional[data-wb-randomize]" ).each( function() {
+	var $elm = $( this ),
+		command = $elm.data( "wb-randomize" ),
+		shuffle = command.shuffle,
```
Comment (duboisp) [inline src/plugins/wb-randomize/wb-randomize.js]: Please rename the variable "command" for "settings".

File: src/plugins/wb-randomize/wb-randomize.js [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+		selector = command.selector,
+		number = command.number,
+		context = $elm,
```
Comment (duboisp) [inline src/plugins/wb-randomize/wb-randomize.js]: "context" variable is not used

File: src/plugins/wb-randomize/wb-randomize.js [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+		selected = $( selector, context );
+
+	selected.shuffle = wb.shuffle;
+
+	if ( shuffle ) {
+		selected = selected.shuffle();
+	}
+
+	selected.rand = wb.rand;
+	selected.rand( number ).toggleClass( "hide" );
```
Comment (duboisp) [inline src/plugins/wb-randomize/wb-randomize.js]: It always toggle? I though it was either shuffle or toggle or "shuffle and toggle"

File: src/plugins/wb-randomize/wb-randomize.js [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+		context = $elm,
+		selected = $( selector, context );
+
+	selected.shuffle = wb.shuffle;
+
+	if ( shuffle ) {
+		selected = selected.shuffle();
+	}
+
+	selected.rand = wb.rand;
```
Comment (duboisp) [inline src/plugins/wb-randomize/wb-randomize.js]: I don't think we need that association

**Review by duboisp — CHANGES_REQUESTED** [review]

File: src/plugins/wb-randomize/wb-randomize.js (hunk 4 above)
Comment (namjohn920) [reply src/plugins/wb-randomize/wb-randomize.js]: context is used in line 18

File: src/plugins/wb-randomize/wb-randomize.js (hunk 6 above)
Comment (namjohn920) [reply src/plugins/wb-randomize/wb-randomize.js]: without this, random function would not work.

File: src/plugins/wb-randomize/wb-randomize.js (hunk 5 above)
Comment (namjohn920) [reply src/plugins/wb-randomize/wb-randomize.js]: so what shuffle does is to shuffle the order of child components. For example, if I have img1, img2 and img3 in order and if I use shuffle, the order will be mixed. What toggle does is to select a certain number of child elements.

Comment (namjohn920) [reply src/plugins/wb-randomize/wb-randomize.js]: I guess if it were to only mix the order, toggle should not be default. Please let me know what your opinion is. I will fix accordingly.

File: site/pages/docs/ref/provisional-en.hbs [hunk 7]
```diff
@@ -111,3 +117,25 @@
...
 	&lt;/div&gt;
 &lt;/div&gt;</code></pre>
+
+<h3 id="well-header-rwd-bgimg">Showing random image using [data-wb-randomize] plugin</h3>
+<p>Below example shows one image out of three images randomly. Refresh the page to see that the image changes.</p>
+<p>There are three inputs required for this plugin.</p>
+<p><code>shuffle</code> is a boolean value and if it is true, it will shuffle the order of child components (in this example, the child components are three <code>img</code>s).</p>
+<p><code>toggle</code> is a boolean value and if it is true, it will toggle 'hide' class of child components to reveal (in this example, the child components are three <code>img</code>s).</p>
+<p><code>selector</code> specifies the type of child components to be displayed.</p>
+<p><code>number</code> specifies the number of child components to be displayed. If it is more than the number of child components, all of them will be displayed.</p>
```
Comment (duboisp) [inline site/pages/docs/ref/provisional-en.hbs]: Please use a definition list instead of paragraph

File: theme/_provisional.scss [hunk 8]
```diff
@@ -2,4 +2,8 @@
 	&.bg-cover, .bg-cover {
 		background-size:cover;
 	}
+
+	.provisional-img-example {
+		width: 100%;
```
Comment (duboisp) [inline theme/_provisional.scss]: I don't think that CSS is needed. There is a generic CSS class name called `full-width` or `img-responsive`. Please remove and update the working example

**Review by duboisp — CHANGES_REQUESTED** [review]

File: src/plugins/wb-randomize/wb-randomize.js [hunk 9]
```diff
@@ -0,0 +1,38 @@
...
+			setting = $elm.data( "wb-randomize" ),
+			selected = $( setting.selector, $( elm ) );
+
+		if ( setting.shuffle ) {
+			selected = wb.shuffleDOM( selected );
+		}
+
+		if ( setting.toggle ) {
+			selected = wb.pickElements( selected, setting.number );
+			selected.toggleClass( "hide" );
```
Comment (duboisp) [inline src/plugins/wb-randomize/wb-randomize.js]: Put this class as a setting, and set in your default props

**Review by duboisp — CHANGES_REQUESTED** [review]

File: site/pages/docs/ref/provisional-fr.hbs [hunk 10]
```diff
@@ -113,3 +119,27 @@
...
 </div>
+
+<h3 id="well-header-rwd-bgimg">Showing random image using [data-wb-randomize] plugin</h3>
+<p>Below example shows one image out of three images randomly. Refresh the page to see that the image changes.</p>
+<dl>
+	<dt><code>shuffle</code> is a boolean value and if it is true, it will shuffle the order of child components (in this example, the child components are three <code>img</code>s).</dt>
+	<dt><code>selector</code> specifies the type of child components to be displayed.</dt>
+	<dt><code>number</code> specifies the number of child components to be displayed. If it is more than the number of child components, all of them will be displayed.</dt>
+	<dt><code>toggle</code> is a boolean value and if it is true, it will toggle 'hide' class of child components to reveal (in this example, the child components are three <code>img</code>s).</dt>
+	<dt><code>toggleClass</code> specifies the class that will be toggled.</dt>
```
Comment (duboisp) [inline site/pages/docs/ref/provisional-fr.hbs]: Merge toggle and toggleClass togetter,

So the toggle setting will contains the CSS class to toggle,

File: src/plugins/wb-randomize/wb-randomize.js [hunk 11]
```diff
@@ -0,0 +1,38 @@
...
+ * @license wet-boew.github.io/wet-boew/License-en.html / wet-boew.github.io/wet-boew/Licence-fr.html
+ * @author @masterbee
+ */
+( function( $, wb ) {
+"use strict";
+
+var $document = wb.doc,
+	componentName = "wb-randomize",
+	selector = ".provisional[data-wb-randomize]",
+	initEvent = "wb-init" + selector,
```
Comment (duboisp) [inline src/plugins/wb-randomize/wb-randomize.js]: Add a default JSO for the "number" setting, like "1"

File: src/plugins/wb-randomize/wb-randomize.js [hunk 12]
```diff
@@ -0,0 +1,38 @@
...
+	init = function( event ) {
+		var elm = wb.init( event, componentName, selector ),
+			$elm = $( elm ),
+			setting = $elm.data( "wb-randomize" ),
+			selected = $( setting.selector, $( elm ) );
+
```
Comment (duboisp) [inline src/plugins/wb-randomize/wb-randomize.js]: Here add a conditional that selected must be valid and contains items.
Throw an error if it is not provided or return nothing

File: site/pages/docs/ref/provisional-fr.hbs [hunk 13]
```diff
@@ -113,3 +119,27 @@
 	&lt;/div&gt;
 &lt;/div&gt;</code></pre>
 </div>
+
+<h3 id="well-header-rwd-bgimg">Showing random image using [data-wb-randomize] plugin</h3>
+<p>Below example shows one image out of three images randomly. Refresh the page to see that the image changes.</p>
+<dl>
+	<dt><code>shuffle</code> is a boolean value and if it is true, it will shuffle the order of child components (in this example, the child components are three <code>img</code>s).</dt>
+	<dt><code>selector</code> specifies the type of child components to be displayed.</dt>
```
Comment (duboisp) [inline site/pages/docs/ref/provisional-fr.hbs]: Add the mention this property is required.

**Review by duboisp — CHANGES_REQUESTED** [review]

File: src/core/helpers.js [hunk 14]
```diff
@@ -72,6 +72,60 @@ wb.download = function( blob, filename, title ) {
...
+	}
+
+	return $( shuffled );
+};
+
+/* ---------------------------------
+@extension: pickElements
+@returns: [collection] of randoms elements
+-------------------------------- */
+wb.pickElements = function( DOM, k ) {
```
Comment (duboisp) [inline src/core/helpers.js:100]: Can you try to unminify the code in this function? Like try to put more descriptive variable name.

Like changing `DOM` for `elm` for consistency with the rest of WET code that make reference to an DOM element or for `$elm` if `DOM` is a jQuery object.

This assignation have no signifiant value: `b = DOM`

File: src/plugins/wb-randomize/wb-randomize.js [hunk 15]
```diff
@@ -0,0 +1,43 @@
...
+		if ( selected.length === 0 || !setting.selector ) {
+			console.error( "selector attribute in data-wb-randomize plugin is undefined or not valid" );
```
Comment (duboisp) [inline src/plugins/wb-randomize/wb-randomize.js:21]: Change `console.error` for `throw`

File: src/plugins/wb-randomize/wb-randomize.js [hunk 16]
```diff
@@ -0,0 +1,43 @@
...
+
+var $document = wb.doc,
+	componentName = "wb-randomize",
+	selector = ".provisional[data-wb-randomize]",
+	initEvent = "wb-init" + selector,
+	init = function( event ) {
+		var elm = wb.init( event, componentName, selector ),
+			$elm = $( elm ),
+			setting = $elm.data( "wb-randomize" ),
+			selected = $( setting.selector, $( elm ) );
```
Comment (duboisp) [inline src/plugins/wb-randomize/wb-randomize.js:18]: Change `$( elm )` for `$elm`

File: src/plugins/wb-randomize/wb-randomize.js [hunk 17]
```diff
@@ -0,0 +1,43 @@
...
+		if ( setting.shuffle ) {
+			selected = wb.shuffleDOM( selected );
+		}
+
+		if ( setting.toggle ) {
+			selected = wb.pickElements( selected, setting.number ? setting.number : 1  );
+			selected.toggleClass( setting.toggle );
+		}
+
+		wb.ready( $( elm ), componentName );
```
Comment (duboisp) [inline src/plugins/wb-randomize/wb-randomize.js:34]: Change `$( elm )` for `$elm`

**Review by duboisp — CHANGES_REQUESTED** [review]

File: site/pages/docs/ref/provisional-en.hbs [hunk 18]
```diff
@@ -111,3 +117,27 @@
 		&lt;p&gt;Secondary title&lt;/p&gt;
 	&lt;/div&gt;
 &lt;/div&gt;</code></pre>
+
+<h3 id="well-header-rwd-bgimg">Showing random image using [data-wb-randomize] plugin</h3>
+<p>Below example shows one image out of three images randomly. Refresh the page to see that the image changes.</p>
+<p>There are three inputs required for this plugin.</p>
+<dl>
+	<dt><code>selector</code> is a required attribute. This attribute specifies the type of child components to be displayed.</dt>
```
Comment (duboisp) [inline site/pages/docs/ref/provisional-en.hbs:125]: closing "</dt>" before the "is". And then wrap the rest with "<dd>" element

File: site/pages/docs/ref/provisional-en.hbs [hunk 19]
```diff
@@ -111,3 +117,27 @@
...
+	<dt><code>shuffle</code> is a boolean value and if it is true, it will shuffle the order of child components (in this example, the child components are three <code>img</code>s).</dt>
+	<dt><code>toggle</code> specifies the class that will be toggled in child components and it will toggle the specified class of child components to reveal (in this example, the name of class that will be toggled is 'hide' in three <code>img</code>s).</dt>
+	<dt><code>number</code> specifies the number of child components to be displayed when toggled. If it is more than the number of child components, all of them will be displayed. By default, it is set to 1.</dt>
+</dt>
```
Comment (duboisp) [inline site/pages/docs/ref/provisional-en.hbs:129]: should be a closing "dl"

**Review by duboisp — CHANGES_REQUESTED** [review]

**Pull request closed without merging**